cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -800,34 +773,23 @@ private void runWithProgramLogging(ProgramRunId programRunId, Map<String, String
     if (future.cancel(true)) {
       LOG.debug("Cancelled {} task for program run {}.", taskKey.getType(), taskKey.getProgramRunId());
       // this is the task state after it has been cancelled
-      ProvisioningTaskInfo currentTaskInfo = Transactionals.execute(transactional, dsContext -> {
-        ProvisionerStore provisionerStore = ProvisionerStore.get(dsContext, datasetFramework);
-
-        ProvisioningTaskInfo currentInfo = provisionerStore.getTaskInfo(taskKey);
-        if (currentInfo == null) {
-          return null;
-        }
-
-        // write that the state has been cancelled. This is in case CDAP dies or is killed before the cluster can
-        // be deprovisioned and the task state cleaned up. When CDAP starts back up, it will see that the task is
-        // cancelled and will not resume the task.
-        ProvisioningOp newOp =
-          new ProvisioningOp(currentInfo.getProvisioningOp().getType(), ProvisioningOp.Status.CANCELLED);
-        ProvisioningTaskInfo newTaskInfo = new ProvisioningTaskInfo(currentInfo, newOp, currentInfo.getCluster());
-        provisionerStore.putTaskInfo(newTaskInfo);
-        LOG.trace("Recorded cancelled state for {} task for program run {}.",
-                  taskKey.getType(), taskKey.getProgramRunId());
-
-        return currentInfo;
-      });
-
+      ProvisioningTaskInfo currentTaskInfo = provisionerStore.getTaskInfo(taskKey);
       // if the task state was gone by the time we cancelled, it's like we did not cancel it, since it's already
       // effectively done.
       if (currentTaskInfo == null) {
         return Optional.empty();
       }
-
-      // run any action that should happen if it was cancelled, like writing a message to TMS
+      // write that the state has been cancelled. This is in case CDAP dies or is killed before the cluster can
+      // be deprovisioned and the task state cleaned up. When CDAP starts back up, it will see that the task is
+      // cancelled and will not resume the task.
+      ProvisioningOp newOp =
+        new ProvisioningOp(currentTaskInfo.getProvisioningOp().getType(), ProvisioningOp.Status.CANCELLED);
+      ProvisioningTaskInfo newTaskInfo = new ProvisioningTaskInfo(currentTaskInfo, newOp, currentTaskInfo.getCluster());
+      provisionerStore.putTaskInfo(newTaskInfo);
```

Comment:
this was previously in a single transaction. I think that it would be better to add a method to the store that updates the state of a task, and returns the previous task info.

Reply:
Added a  TODO: CDAP-14909 since we have to revisit transactions after all the relevant store in provisioning service is migrated. Leaving it as two tx for now.

Reply:
I don't see why we need to revisit this later. I only see a single dataset/table involved in this.

Reply:
Updated the provisioner store with the single tx as per our discussion